---
title: "scMapping(single-cell mapping)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Performing cell type annotation in single-cell RNA-sequencing data analysis}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


The goal of scMapping is to annotate cell clusters for single cell data.

# 1. Installation

You can install the development version of scMapping like so:

``` {r, message=FALSE, warnings = F, results=FALSE}
# devtools::install_github("chloelulu/scMapping")
# pkgs <- c("ggplot2", "dplyr","tidyr","tibble","reshape2")
# new.packages <- pkgs[!(pkgs %in% installed.packages()[,"Package"])]
# if(length(new.packages)) install.packages(new.packages)
```

Library required packages
```{r, message=FALSE, warning=FALSE}
pkgs <- c("ggplot2", "dplyr","tidyr","tibble","reshape2")
sapply(pkgs, require, character.only = TRUE)
library(scMapping)
```

After you load the package, the demo “data” and integrated “scMappingDatabase”
will be automatically loaded

``` {r, message=FALSE, warning=FALSE}
head(data)
```

scMappingDatabase is the database integrated in the package. User can also supply their own database, should be in the same format as scMappingDatabase

``` {r, message=FALSE, warning=FALSE}
head(scMappingDatabase[,c(1:4)])
```


# 2. Running scMapping

To annotate a scRNA-seq data analysis output generated by ‘FindAllMarkers’ function of
Seurat[(Butler, A., et al.)](https://www.nature.com/articles/nbt.4096?report=reader)


```{r, message=FALSE, warning=FALSE}
obj <- scmapping(data = data, database = scMappingDatabase, tissue = 'muscle')
```


scMapping also supplies a plot function to help you visualize the result. If "directory = NULL" plot will be automatically saved to the current working directory.
```{r,fig.retina = 4, fig.width= 7, fig.height=5, results=FALSE, message=FALSE, warning=FALSE}
plt <- scmapping.plot(scmapping.object = obj, directory = NULL)
```

# 3. Using scMapping start from Seurat 

Below code illustrate how to use scMapping if you start with a Seurat analysis.
```{r,fig.retina = 4, fig.width= 7, fig.height=5, results=FALSE, message=FALSE, warning=FALSE}
library(SeuratData)
library(Seurat)
seurat.obj <- LoadData(ds = "pbmc3k") 
seurat.obj <- NormalizeData(object = seurat.obj)
seurat.obj <- FindVariableFeatures(object = seurat.obj)
seurat.obj <- ScaleData(object = seurat.obj)
seurat.obj <- RunPCA(object = seurat.obj)
seurat.obj <- FindNeighbors(object = seurat.obj)
seurat.obj <- FindClusters(object = seurat.obj)
seurat.markers <- FindAllMarkers(seurat.obj, method = 'MAST')
scmapping.obj <- scmapping(data = seurat.markers, database=scMappingDatabase, tissue = 'blood')
plt <- scmapping.plot(scmapping.object = scmapping.obj, directory = NULL)
```

You can also extract the markers mapped by scMapping and making a dotplot. 
Below code visualize the genes mapped to cluster 0 (predicted cell type Granulocyte)
```{r,fig.retina = 4, fig.width= 8, fig.height=5, results=FALSE, message=FALSE, warning=FALSE}
gns <- scmapping.obj$markers$genes[scmapping.obj$markers$cluster==0 & scmapping.obj$markers$celltype=='Granulocyte']
gns <- strsplit(gns, ',')[[1]]
DotPlot(seurat.obj, features = gns)
```

# 4. References

  - Lu Yang, Yan Er Ng, Haipeng Sun, Ying Li, Nathan LeBrasseur, Jun Chen, Xu Zhang. single-cell mapping (scMapping), an easy-to-use tool for mapping cell types in single-cell RNA-sequencing data analysis. In preparation



# 5. Session info
```{r}
sessionInfo()
```



